packages/nx/src/project-graph/utils/project-configuration-utils.ts

```diff
@@ -554,6 +543,90 @@ export function readProjectConfigurationsFromRootMap(
   return projects;
 }
 
+function validateAndNormalizeProjectRootMap(
+  projectRootMap: Record<string, ProjectConfiguration>
+) {
+  // Name -> Project, used to validate that all projects have unique names
+  const projects: Record<string, ProjectConfiguration> = {};
+  // If there are projects that have the same name, that is an error.
+  // This object tracks name -> (all roots of projects with that name)
+  // to provide better error messaging.
+  const conflicts = new Map<string, string[]>();
+  const projectRootsWithNoName: string[] = [];
+
+  for (const root in projectRootMap) {
+    const project = projectRootMap[root];
+    // We're setting `// targets` as a comment `targets` is empty due to Project Crystal.
+    // Strip it before returning configuration for usage.
+    if (project['// targets']) delete project['// targets'];
+
+    try {
+      validateProject(project, projects);
+    } catch (e) {
+      if (isProjectsWithNoNameError(e)) {
+        projectRootsWithNoName.push(...e.projectRoots);
+      } else if (isProjectsWithConflictingNamesError(e)) {
+        const newConflicts = e.conflicts;
+        for (const [name, roots] of newConflicts) {
+          const conflictsArray = conflicts.get(name) ?? [];
+          conflictsArray.push(...roots);
+          conflicts.set(name, conflictsArray);
+        }
+      }
+    }
+
+    for (const targetName in project.targets) {
+      project.targets[targetName] = normalizeTarget(
+        project.targets[targetName],
+        project
+      );
+
+      if (
+        !project.targets[targetName].executor &&
+        !project.targets[targetName].command
+      ) {
+        delete project.targets[targetName];
+      }
+    }
+  }
+
+  if (conflicts.size > 0) {
+    throw new ProjectsWithConflictingNamesError(conflicts, projects);
+  }
+  if (projectRootsWithNoName.length > 0) {
+    throw new ProjectsWithNoNameError(projectRootsWithNoName, projects);
+  }
+  return projectRootMap;
+}
+
+export function validateProject(
+  project: ProjectConfiguration,
+  knownProjects: Record<string, ProjectConfiguration>
+) {
+  if (!project.name) {
+    try {
+      const { name } = readJsonFile(join(project.root, 'package.json'));
+      if (!name) {
+        throw new Error(`Project at ${project.root} has no name provided.`);
+      }
+      project.name = name;
+    } catch {
+      throw new ProjectsWithNoNameError([project.root], knownProjects);
+    }
+  }
+  if (
+    knownProjects[project.name] &&
+    knownProjects[project.name].root !== project.root
+  ) {
+    throw new ProjectsWithConflictingNamesError(
```

Comment:
Make this a different type or a single project which has a conflicting name.. with a previous project.